### PR TITLE
KAFKA-20397 Avoid blocking on fetchBuffer when metadata error is pending

### PR DESCRIPTION
**Pull request by nileshkumar3**
When a metadata error (e.g. TopicAuthorizationException) is discovered by the background thread
and set on the inflight AsyncPollEvent between checkInflightPoll() and fetchBuffer.awaitWakeup()
in pollForFetches(), the consumer blocks for the full poll timeout before surfacing the error.
This adds a check in pollForFetches() to skip the blocking wait when the inflight poll event
already has a metadata error, allowing the error to be surfaced promptly on the next iteration
of the poll loop via checkInflightPoll().

Reviewers - @lianetm @AndrewJSchofield @kirktrue

Reviewers: Sushant Mahajan <smahajan@confluent.io>

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -1991,6 +1990,13 @@ private Fetch<K, V> pollForFetches(Timer timer) {
             }
         }
 
+        // If the background thread has already discovered a metadata error (e.g. TopicAuthorizationException)
+        // and completed the inflight poll event, skip blocking on the fetch buffer entirely. The error will
+        // be surfaced by checkInflightPoll on the next iteration of the poll loop. See KAFKA-20397.
+        if (inflightPoll != null && inflightPoll.error().isPresent()) {
+            return collectFetch();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:2002]: What's the purpose behind calling `collectFetch()`? There won't be anything returned will there?

I'm thinking about the corner case where we _don't_ loop again and call `checkInflightPoll()` because the timer is exhausted. I guess in that case we'll hit the error the next time the user calls `poll()` 🤔

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 2]
```diff
@@ -1736,6 +1737,73 @@ public void testPollDoesNotAddNewAsyncPollEventWhenOneIsAlreadyInFlight() {
...
+        // This models the background thread completing the event with a metadata error
+        // between checkInflightPoll and the blocking wait.
+        doAnswer(invocation -> {
+            if (!errorInjected.get() && capturedEvent.get() != null) {
+                capturedEvent.get().onMetadataError(
+                    new TopicAuthorizationException(singleton(topicName)));
+                errorInjected.set(true);
+            }
+            return fetchWaitMs;
+        }).when(applicationEventHandler).maximumTimeToWait();
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:1872]: Could this be simplified:

```suggestion
        doAnswer(invocation -> {
            AsyncPollEvent argument = invocation.getArgument(0);
            argument.onMetadataError(
                new TopicAuthorizationException(singleton(topicName)));
            return null;
        }).when(applicationEventHandler).add(ArgumentMatchers.isA(AsyncPollEvent.class));
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java [hunk 3]
```diff
@@ -1736,6 +1737,73 @@ public void testPollDoesNotAddNewAsyncPollEventWhenOneIsAlreadyInFlight() {
...
+
+        doAnswer(invocation -> {
+            Timer pollTimer = invocation.getArgument(0, Timer.class);
+            ((MockTime) time).sleep(pollTimer.remainingMs());
```
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: `time` doesn't need to be cast as `MockTime` to use `sleep()`.

**Review by kirktrue — COMMENTED** [review]
Just a quick review.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 1 above)
Comment (nileshkumar3) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:2002]: Good point. collectFetch() will almost always return empty here — the buffer was already drained at the top of pollForFetches. The only edge case is if position validation completed between line 1965 and 1996, making previously-skipped buffer data readable. In that case, returning the data seems fine since it was legitimately fetched, and the error will be thrown on the next poll().

The real goal is just to skip awaitWakeup(), which works either way. I am fine  to change to Fetch.empty() if you think that make more sense.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java (hunk 3 above)
Comment (nileshkumar3) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: done.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java (hunk 2 above)
Comment (nileshkumar3) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java:1872]: Took your suggestion — removed the AtomicBoolean guard and the MockTime cast, thanks!

On injecting in add() vs maximumTimeToWait(): if we inject in add(), the error gets caught inside checkInflightPoll() via maybeClearCurrentInflightPoll() — that's the existing error path. The test needs the error to arrive after checkInflightPoll() but before awaitWakeup(), which is the race condition this fix addresses. maximumTimeToWait() gives us that timing. The AtomicReference is just needed because maximumTimeToWait() has no arguments to extract the event from.

**Comment by nileshkumar3** [comment]
> Just a quick review.



> Just a quick review.

Thank you for the Review.

**Comment by nileshkumar3** [comment]
@kirktrue All comment addressed, please have another look when you get a chance.

**Comment by nileshkumar3** [comment]
@kirktrue  Gentle ping for the review.

**Comment by nileshkumar3** [comment]
@kirktrue  @lianetm Could you please review this when you get a chance.

**Comment by nileshkumar3** [comment]
@kirktrue Gentle Ping to review this.

**Comment by nileshkumar3** [comment]
@lianetm  Could you please review when you get a chance.

**Review by smjn — APPROVED** [review]
Thanks for the changes, LGTM

**Comment by nileshkumar3** [comment]
Thanks @smjn for the review.

**Comment by nileshkumar3** [comment]
@lianetm gentle ping to review this.

**Comment by nileshkumar3** [comment]
@lianetm gentle ping on this.
